casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala

```diff
@@ -53,34 +61,163 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
-        block1 <- nodes(1).addBlock(deploys(1))
-        _      <- nodes(0).receive()
-        _      <- nodes(1).receive()
-        _      <- nodes(0).receive()
-        _      <- nodes(1).receive()
+    def echoContract(no: Int) =
+      Rho(s"""new stdout(`rho:io:stdout`) in { stdout!("Contract $no") }""")
+    merges(echoContract(1), echoContract(2), Rho("Nil"))
+  }
 
-        //multiparent block joining block0 and block1 since they do not conflict
-        multiparentBlock <- nodes(0).addBlock(deploys(1))
-        _                <- nodes(1).receive()
+  // TODO: Peek rows/column
+  // Note this skips pairs that lead to infinite loops
+  it should "handle multi-parent blocks correctly when they operate on volatile produce/consume pairs" in effectTest {
+    // Sends (linear sends)
+    val S0 = Rho("@0!(0)")
+    val S1 = Rho("@0!(1)")
+    // Repeats (persistent sends)
+    val R0 = Rho("@0!!(0)")
+    val R1 = Rho("@0!!(1)")
+    // For-s (linear receives)
+    val F_ = Rho("for (_ <- @0) { 0 }")
+    val F0 = Rho("for (@0 <- @0) { 0 }")
+    val F1 = Rho("for (@1 <- @0) { 0 }")
+    // Contracts (persistent receives)
+    val C_ = Rho("contract @0(id) = { 0 }")
+    val C0 = Rho("contract @0(@0) = { 0 }")
+    val C1 = Rho("contract @0(@1) = { 0 }")
+    //FIXME add missing cases for in-deploy COMM-s wherever there's a pair without X (a COMM)
+    //FIXME all `conflictsForNow` should eventually be replaced with `merges`
+    Map(
+      "!X !X"        -> conflictsForNow(S0, S0, Nil),
+      "!X !4"        -> conflictsForNow(S0, S0, F_),
+      "!X (!4)"      -> merges(S0, S1 | F_, Nil),
+      "!X !C"        -> conflictsForNow(S0, S1, C1),
+      "!X (!C)"      -> conflicts(S0, S1 | C_, Nil),
+      "!X 4X"        -> conflicts(S0, F_, Nil),
+      "!X 4!"        -> conflictsForNow(S0, F_, S0),
+      "!X (4!)"      -> coveredBy("!X (!4)"),
+      "!X 4!!"       -> conflictsForNow(S0, F_, R0),
+      "!X (4!!)"     -> conflictsForNow(S0, F_ | R0, Nil),
+      "!X !!X"       -> conflictsForNow(S0, R0, Nil),
+      "!X !!4"       -> conflictsForNow(S0, R1, F1),
+      "!X (!!4)"     -> coveredBy("!X (4!!)"),
+      "!X CX"        -> conflicts(S0, C_, Nil),
+      "!X C!"        -> conflicts(S0, C_, S0),
+      "!X (C!)"      -> conflicts(S0, C_ | S0, Nil),
+      "!4 !4 same 4" -> conflicts(S0, S1, F_),
+      "!4 !4 diff 4" -> conflicts(S0, S1, F0 | F1),
+      "!4 (!4)"      -> merges(S0, S1 | F1, F0),
+      "(!4) (!4)"    -> merges(S0 | F_, S0 | F_, Nil),
+      "!4 !C"        -> conflictsForNow(S0, S1, F0 | C1),
+      "!4 4X"        -> conflictsForNow(S0, F_, F_),
+      "!4 4!"        -> conflictsForNow(S0, F_, F0 | S1),
+      "!4 4!!"       -> conflictsForNow(S0, F_, F0 | R1),
+      "!4 !!X"       -> conflictsForNow(S0, R1, F0),
+      "!4 !!4"       -> conflictsForNow(S0, R1, F_ | F1),
+      "!4 CX"        -> conflictsForNow(S0, C_, F_),
+      "!4 C!"        -> conflictsForNow(S0, C_, F0 | S1),
```

Comment:
shouldn't this `C_` be `C1`?

Reply:
I was under the impression that we want to have a set of "rules" that describe how merging works.  the `!4 C!` has a very specific meaning. introducing generality in the implementation obfuscates the meaning. note that there is no generality in the term. it is a description of state.
IOW `!4 C!` is not the same as (S0, C_, F0 | S1). 
I understand that replacing the `_` will not yield equality but it will bring the implementation closer to what it should hold.

IF we had means to express the left hand in a better way than by discriminating with patterns - maybe it would be ok to try and write general laws. but we don't.

this is my 0.02$, do with it as you please.

Reply:
Notice we're talking about a test, not an implementation. The more general the tests, the more constrained the tested implementation.

Take for example `merges(S0, C_, F_)`.
Replacing it with `merges(S0, C0, F0)` poses the risk that the tested code will  only merge when the receives have the exact same matching pattern.

Granted, one could argue that `merges(S0, C_, F_)` will not spot cases where `merges(S0, C_, F0)` or `merges(S0, C0, F_)` would not hold. That would though require an odd explicit condition in the prod code, and can be fought by adding additional cases for specific data.

In general, I believe it's good to keep the test data minimal (C_ < C0) - which helps a better implementation, less over-specifying tests, and stays as close to the actual rules of the tested domain as possible.